### PR TITLE
[go_router_builder] Support default value for `Set`, `List` and `Iterable` route parameters

### PR DESCRIPTION
**Pull request by ValentinVignal**
Fixes: https://github.com/flutter/flutter/issues/122003

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Review by chunhtai — COMMENTED** [review]
The Code LGTM, can you add a test in example/test to make sure thing works as expected if navigate to the route with default value?

**Comment by ValentinVignal** [comment]
@chunhtai I added a test in [test: Add a test in example with default values](https://github.com/flutter/packages/pull/3391/commits/4b422b660cd87b00db5740d5477bb4a577d58194) but I'm very proud of it. Do you have a better idea?

File: packages/go_router_builder/example/test/all_types_test.dart [hunk 1]
```diff
@@ -136,4 +136,24 @@ void main() {
             '/iterable-route?int-list-field=1&int-list-field=2&int-list-field=3'),
         findsOneWidget);
   });
+
+  testWidgets(
+      'It should navigate to the iterable route with the its default values',
```
Comment (chunhtai) [inline packages/go_router_builder/example/test/all_types_test.dart:141]:
```suggestion
      'It should navigate to the iterable route with its default values',
```

File: packages/go_router_builder/example/lib/all_types.dart [hunk 2]
```diff
@@ -342,6 +344,77 @@ class IterableRoute extends GoRouteData {
...
+  final Iterable<bool> boolIterableField;
+  final List<bool> boolListField;
+  final Set<bool> boolSetField;
+
+  final Iterable<SportDetails> enumIterableField;
+  final List<SportDetails> enumListField;
+  final Set<SportDetails> enumSetField;
+
+  @override
+  Widget build(BuildContext context, GoRouterState state) => BasePage<String>(
```
Comment (chunhtai) [inline packages/go_router_builder/example/lib/all_types.dart:390]: Maybe build a widget that holds all these parameter as its property, and then you can use `tester.widget<>(find.byType)` to get the widget and then you can compare the value of each property in the widget.

Comment (ValentinVignal) [reply packages/go_router_builder/example/lib/all_types.dart]: I've implemented it in [test: Add better tests](https://github.com/flutter/packages/pull/3391/commits/cbfef9d67901501a27c043379d68e209640195c5)

**Review by chunhtai — APPROVED** [review]
LGTM

**Comment by ValentinVignal** [comment]
@hangyujin Is there anything I should do on this PR?

**Review by hangyujin — APPROVED** [review]
LGTM

**Pull request merged**